cilium-health: Move to the separate container and systemd service

vadorovsky opened the pull request:
To make work on latency metrics easier, this PR moves cilium-daemon to the separate container (or systemd service). Cilium-health daemon is going to be managed by container orchestration frameworks or systemd, not by cilium-agent.

Ref #4267

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5713)
<!-- Reviewable:end -->

raybejjani commented on Documentation/install/guides/from_source.rst:
```diff
@@ -62,5 +62,7 @@ You can also add it in your ``~/.bashrc`` file:
-    sudo mkdir -p /etc/sysconfig/cilium && cp contrib/systemd/cilium /etc/sysconfig/cilium
+    sudo mkdir -p /etc/sysconfig
+    sudo cp contrib/systemd/cilium /etc/sysconfig/cilium
+    sudo cp contrib/systemd/cilium-health /etc/sysconfig/cilium
```
the target should be `/etc/sysconfig/cilium-health`, right?

vadorovsky replied:
Yes, good catch

raybejjani commented on cilium-health/launch/launcher.go:
```diff
@@ -1,102 +0,0 @@
-	for {
-		var err error
-
-		os.Remove(defaults.SockPath)
```
I don't see this replicated in the systemd units or in the cilium-health startup. I'm guessing it was there to fix some behaviour at some point. We might need to keep it on cilium-health startup.

vadorovsky replied:
Good catch. I'm not sure what was the motivation behind socket removal. I see that it was introduced there:

https://github.com/cilium/cilium/commit/b18e3ba88e2ffd73acc3ad1a2c7f02152bf513a8

@joestringer Do you think it's fine if I will remove that unix socket in systemd service and yaml files? I guess it's just for cleaning the leftover socket after previous run of cilium-health, right?

vadorovsky replied:
nvm

I realized that this socket removal, as well as waiting for Cilium API, is needed for graceful start of `cilium-health -d`. I moved those pieces to `cilium-health/cmd/root.go`.

vadorovsky replied:
Please note that I added maximum number of retries to connect to Cilium API:

https://github.com/cilium/cilium/blob/287bbdb29a0264c352caddc887e5931829903faa/cilium-health/cmd/root.go#L156

I think that 2 minutes are fair as a max timeout and I think that cilium-health daemon/container should be failing if Cilium API is not available long-term.

raybejjani commented on contrib/systemd/cilium-health.service:
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=cilium-health
+Documentation=http://docs.cilium.io
+Requires=cilium.service
```
I don't think this is true, but did you check that cilium doesn't do something funky if cilium-health isn't up? I think the answer is no, and that it's fine.

vadorovsky replied:
Good question, I will check today.

vadorovsky replied:
The answer is no. :) Cilium daemon seems to work file after stopping/disabling cilium-health.

raybejjani commented on daemon/status.go:
```diff
@@ -162,10 +162,6 @@ func (d *Daemon) getStatus() models.StatusResponse {
 	sr.Cluster = d.getNodeStatus()
 
-	if d.ciliumHealth != nil {
-		sr.Cluster.CiliumHealth = d.ciliumHealth.GetStatus()
```
It might still be worthwhile to keep this status. It would mean keeping/copying the health client stuff somewhere.

joestringer replied:
Seems reasonable. I'm not sure whether we want `cilium status` to fail if this fails, though (doesn't really make sense in k8s environments since `cilium-health` would have its own health checks, then in non-k8s environments someone may choose just not to run `cilium-health`)

raybejjani replied:
True... I guess this also means we might need to change docs that talk about cilium-health.

vadorovsky replied:
So let me understand your suggestion. You want to:

- keep the information about `cilium-health` in `cilium status` (just to make it easier for people using Cilium with systemd, without Kubernetes)
- but return status 0 even if `cilium-health status` fails

Is that correct?

joestringer replied:
> return status 0 even if `cilium-health status` fails

To clarify: return status 0 from `cilium status` if its call to `cilium-health status` fails. The output should still print that it was unable to connect to `cilium-health` with a user-friendly string describing that this may or may not represent a real issue.

vadorovsky replied:
`cilium status` already returns 0 even if cilium health is unreachable, there is no need to change code.

```
root@runtime1:~# systemctl stop cilium-health
root@runtime1:~# cilium status
KVStore:                Ok         Consul: 172.17.0.3:8300
ContainerRuntime:       Ok         docker daemon: OK
Kubernetes:             Disabled   
Cilium:                 Ok         OK
NodeMonitor:            Disabled
Cilium health daemon:   Warning   Get http:///var/run/cilium/health.sock/v1beta/hello: dial unix /var/run/cilium/health.sock: connect: no such file or directory
IPv4 address pool:      3/65535 allocated
IPv6 address pool:      2/65535 allocated
Controller Status:      13/13 healthy
Proxy Status:           OK, ip 10.11.0.1, port-range 10000-20000
Cluster health:         Warning   cilium-health daemon unreachable
root@runtime1:~# echo $?
0
```

raybejjani reviewed (changes requested):
I'm mostly reviewing this as a Janitor, so I might have missed some discussion in the other PR that became this one (among others). This seems fine to me overall, with the exception of losing the cilium-health status (maybe? perhaps the one we have actually comes from the other health component).

joestringer commented on Documentation/contributing.rst:
```diff
@@ -162,7 +162,9 @@ directly and manually install Cilium:
-        $ sudo cp contrib/systemd/cilium  /etc/sysconfig/cilium
+        $ sudo cp contrib/systemd/cilium-health.service /etc/systemd/system/
+        $ sudo cp contrib/systemd/cilium /etc/sysconfig/cilium
+        $ sudo cp contrib/systemd/cilium-health /etc/sysconfig/cilium
```
Same comment as @raybejjani made here too.

joestringer replied:
Also, looks like there's a `systemctl restart cilium` below which doesn't have a corresponding setup for the cilium-health daemon. Maybe this is deliberate and you intend to put that in other documentation, in which case that's fine too.

vadorovsky replied:
Fixed, I added a `restart` command for `cilium-health`.

joestringer commented on contrib/systemd/cilium-health:
```diff
@@ -0,0 +1,22 @@
+# Run as a daemon:
+# --daemon
+#
+# Note: Check more cli options using cilium-health-h
```
`cilium-health -h` or `cilium-health --help` (missing space)

joestringer commented on contrib/vagrant/start.sh:
```diff
@@ -355,7 +356,8 @@ if [ -n "\${K8S}" ]; then
 
-service cilium restart
+systemctl restart cilium-health
+systemctl restart cilium
```
Reverse the order of these two?

raybejjani replied:
Hmm, since the systemd unit for cilium-health relies on the cilium unit, would restarting cilium restart cilium health too? I don't know what systemd does with it's dependency graph.

vadorovsky replied:
It should, I will check and let you know later. And of course correct the docs.

joestringer commented on examples/kubernetes/templates/v1/cilium-crio-ds.yaml.sed:
```diff
@@ -181,6 +181,33 @@ spec:
+            exec:
+              command:
+              - cilium-health
+              - status
```
You might want to consider using --succinct here too, the intention behind succinct is to print one line and succeed/fail (though I'm not sure it is quite acting as intended just yet.. but that's a matter for later).

vadorovsky replied:
Done, added.

joestringer reviewed (changes requested):
Thanks for splitting this out, glad to see some progress here.

I didn't catch the CI failure and it looks like it's been deleted so I can't provide any hints there.

Please also add `cilium-health` to the `reload` target in `Makefile`.

joestringer commented on examples/kubernetes/templates/v1/cilium-crio-ds.yaml.sed:
```diff
@@ -187,6 +187,35 @@ spec:
               add:
                 - "NET_ADMIN"
             privileged: true
+        - image: docker.io/cilium/cilium:latest
```
The main cilium image on these templates uses` image: docker.io/cilium/cilium:__CILIUM_VERSION__` instead, which presumably picks up the version during template generation. I think we should do the same for these new containers.

joestringer reviewed (approved):
The changes LGTM, just one brief note that *might* be related to the CI failure.

raybejjani reviewed (approved):
Sorry for not approving this earlier! I'm still catching up on email. Thanks for all the work @mrostecki

joestringer commented on cilium-health/cmd/root.go:
```diff
@@ -170,7 +170,7 @@ func runServer() {
 
 	// Wait until Cilium API is available
-	waitForCiliumAPI()
+	// waitForCiliumAPI()
```
FWIW I believe that the intention behind this wait previously was that if you don't wait for the Cilium API to become available, then cilium-health is more likely to complain in the logs about the lack of ability to talk to Cilium (since it will almost immediately start reaching out to cilium to figure out the list of IPs that it should start polling).

vadorovsky replied:
Not only. I also noticed that if you start cilium-health server before cilium-agent, you have two servers running:

```
root@runtime1:~# systemctl status cilium-health
● cilium-health.service - cilium-health
   Loaded: loaded (/lib/systemd/system/cilium-health.service; disabled; vendor preset: enabled)
   Active: active (running) since Wed 2018-12-05 20:00:45 UTC; 3min 44s ago
     Docs: http://docs.cilium.io
 Main PID: 14230 (cilium-health)
    Tasks: 10 (limit: 4684)
   CGroup: /system.slice/cilium-health.service
           └─14230 /usr/bin/cilium-health --daemon

Dec 05 20:00:45 runtime1 systemd[1]: Started cilium-health.
Dec 05 20:00:47 runtime1 cilium-health[14230]: level=info msg="Serving cilium health at http://[::]:4240" subsys=health-server
Dec 05 20:00:49 runtime1 cilium-health[14230]: level=info msg="Serving cilium health at unix:///var/run/cilium/health.sock" subsys=health-server
root@runtime1:~# systemctl cat cilium-health
```

When you uncomment that line, you have only one server running, the one which listenes on U nix socket. That seems strange, I'm trying to debug deeper...

Anyway, I think we should not wait for Cilium API. I guess that waiting for Cilium API was the cause of timeouts in CI.

vadorovsky committed:
doc: Specify container name in Cilium pods

Cilium pod contains two containers, so to avoid warning messages
from kubectl, `kubectl exec` commands should specify cilium-agent
as the container to be executed.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
doc: Add information about cilium-health service

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
vagrant: Add cilium-health service and sysconfig

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
test/helpers: Specify container name in CiliumExec

Cilium pod contains two containers, so to avoid warning messages
from kubectl, cilium-agent container is specified as the container
to be executed.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
cilium-health: Wait for cilium-agent API availability

cilium-health deamon, in order to start properly, needs
availability of cilium-agent API and needs to clean the old
cilium-health socket.

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
cilium-health: daemon: Remove health launcher

cilium-health daemon is running as a separate container or systemd
service, so there is no need to manage it by cilium-agent daemon.

This commit also moves the logic about watching cilium-health
status to cilium-agent daemon.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
examples/kubernetes: Add cilium-health container

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
systemd: Add cilium-health service

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
test: Increase timeout for Cilium readiness

Separated cilium-health requires increased timeout.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>
